c2rust-analyze/src/context.rs

```diff
@@ -84,11 +85,25 @@ pub struct AnalysisCtxt<'a, 'tcx> {
 
     pub local_decls: &'a LocalDecls<'tcx>,
     pub local_tys: IndexVec<Local, LTy<'tcx>>,
+    pub c_void_ptrs: IndexSet<Place<'tcx>>,
     pub addr_of_local: IndexVec<Local, PointerId>,
     /// Types for certain [`Rvalue`]s.  Some `Rvalue`s introduce fresh [`PointerId`]s; to keep
     /// those `PointerId`s consistent, the `Rvalue`'s type must be stored rather than recomputed on
     /// the fly.
     pub rvalue_tys: HashMap<Location, LTy<'tcx>>,
+    /// A mapping for substituting [`Place`]s adhering to the
+    /// following pattern
+    /// ```
+    /// _1 = malloc(...);
+    /// _2 = _1 as *mut T;
+    /// ```
+    ///
+    /// In this case, `_1` would be mapped to `_2`, which is indicative
+    /// of the amended statement:
+    /// ```
+    /// _2 = malloc(...);
+    /// ```
+    pub special_casts: HashMap<Place<'tcx>, Place<'tcx>>,
```

Comment:
Can you newtype `SpecialCasts`?  I think a `struct SpecialCasts(HashMap<Place<'tcx>, Place<'tcx>>)` is better, as that way the intent is a bit clearer, you don't have to repeat the same docs twice, and you can add a few methods that better signify intent and reduce repetition?

Reply:
added in 528ec9e4